### PR TITLE
remove some restrictions on function subtyping

### PR DESCRIPTION
**Pull request by thedufer**
Function subtyping is unnecessarily restrictive - as long as the first unerased argument is positional, erasing optional arguments via type unification is safe, even if there are future labeled arguments. This matches the usual behavior of optional arguments, and allows for functions with both optional arguments and polymorphic return types to be used more easily.  This allows for constructs like (excuse the Base-based example; off the top of my head I don't know a stdlib function with the right properties) `|> Option.value_exn` and `>>| Option.value_exn`.

This change removes the restriction that a function type must contain no labeled arguments in order for type unification to erase optional arguments, but leaves in place the requirement that the first unerased argument must be positional.

I've also added tests showing both the new behavior I'm looking to introduce (the first test) and unifications that must continue to fail (due to the first unerased argument not being positional).

### CODE REVIEW

File: testsuite/tests/typing-func-subtype/opt_arg_subtyping.ml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+let _ = (f : y:'a -> int);;
+
+[%%expect {|
+val f : ?x:'a -> y:'b -> int = <fun>
+Line 4, characters 9-10:
+4 | let _ = (f : y:'a -> int);;
+             ^
+Error: The value "f" has type "?x:'b -> y:'c -> int"
+       but an expression was expected of type "y:'a -> int"
+       Labels "?x" and "y" do not match
```
Comment (smuenzel) [inline testsuite/tests/typing-func-subtype/opt_arg_subtyping.ml:30]: Would be nice if it's possible to say something like this (maybe wording needs to be modified to match the correct constraint)
```suggestion
Error: The value "f" has type "?x:'b -> y:'c -> int"
       but an expression was expected of type "y:'a -> int"
       The optional argument "?x" cannot be erased:
       It is not followed by an unlabeled argument.
```

Or maybe `The last erased argument is not followed by an unlabeled argument.`

Comment (thedufer) [reply testsuite/tests/typing-func-subtype/opt_arg_subtyping.ml:30]: That makes sense, but after some investigation I think it's a more involved change - perhaps a separate PR? The check is currently done in Typecore but the errors are raised in Ctype. If someone wants to give a bit of guidance on the right way to bridge that gap I could take a crack at it but I don't think I currently have enough context on this codebase to take a reasonable approach.

**Comment by garrigue** [comment]
A first remark, there are two notions of function subtyping in the language, in the core language (`subtype_rec`) and in the module language, and this PR is about neither.
It is about function coercion upon application, which is a somehow adhoc feature, but which is careful to preserve some untyped semantics (https://github.com/ocaml/ocaml/pull/10653#issuecomment-1815914118). One would have to check wether this PR preserves this semantics, and also consider principality issues (but IIRC `type_argument` is not strictly principal anyway).

**Comment by thedufer** [comment]
Apologies for the imprecise language - most of my knowledge around OCaml's type system comes from observation of how it behaves, rather than an academic understanding of it, so I've probably gotten a bunch of terminology wrong. I'm more than happy to change the prose here.

My thinking here is that if you can coerce between two types that implies a subtyping relationship between them, but perhaps that's not an appropriate understanding of coercion? I'm not sure application is strictly relevant, unless that means something other than what I expect - the same thing can happen with just a type-hint. Perhaps I should have used that in the tests, but I found the application case a little easier to think about, for whatever reason.

I _believe_ this PR follows the same semantics as optional argument erasure under normal function application. TBH I am having trouble following the reduction semantics described, but I'm somewhat confident about this because the case here is relatively simple compared to the full semantics of function application. Even after this PR, we only allow coercion to occur when a series of optional arguments, all of which are being erased, are immediately followed by a non-labelled argument. Like in trunk, the erased arguments aren't actually applied until that non-labelled argument is.

**Comment by Octachron** [comment]
Sorry for the late analysis but the simplified check is too lax when optional arguments of the coerced function appear in the expected type.
Consider,
```ocaml
let h f = f () ?o:(Some 2) ()
let f ?(o=0) () ?(o=1) () = o
let g () ?o () = f () ?o ()
```
then in untyped OCaml `h f` = `1` whereas `h g = 2` , thus the transformation does not preserve the untyped semantics  in this specific case.

However, if we apply the transformation only if the erased optional arguments do not appear in the expected type, the transformation should be semantically correct under all three semantics of labelled applications (see below for my analysis). 

Indeed let's consider an abstraction `fun ?x₀ … ?xₙ u -> e` used only in a context `f lₖ:vₖ` where $l_k≠ x_s$ then
`fun u -> (fun ?x₀ … ?xₙ u -> e) u` is semantically equivalent to `fun ?x₀ … ?xₙ u -> e`.

### Max-delay semantics

 In this semantics, an application `(fun ?x0 … ?xₙ u -> e) lₖ:vₖ`  ($l_k≠x_s$) is a full-prefix application if and only if
  `(fun u -> (fun ?x0 … ?xₙ u -> e) u) lₖ:vₖ` ($l_k≠x_s$) is a full prefix application: indeed a full-prefix application requires at least the first unlabelled argument, which will then fulfill the optional arguments in both abstractions.

### Strict-eager and unlabelled-strict  
 For these two semantics, if the application does not contain a non-labelled argument then both

 - `(fun x -> (fun ?x0 … ?xₙ u -> e) x)  lₖ:vₖ` with  $l_k≠*$, $l_k≠x_s$ 
-  `(fun ?x0 … ?xₙ u -> e) lₖ:vₖ` $l_k≠*$ with  $l_k≠x_s$

are values.

If the application contains an unlabelled argument, then under both under the strict-eager evaluation and the unlabelled-strict semantics we need to fulfill the optional parameters and evaluate the effects of the corresponding parameters:

In other words, both
- `(fun ?x₀ … ?xₙ u -> e) lₖ:vₖ` with  $l_k≠ x_s, \exists s, l_s = *$
-  `(fun u -> (fun ?x₀ … ?xₙ u -> e) u) lₖ:vₖ` with $l_k≠ x_s, \exists s, l_s = *$
reduce to
- `([vₛ/u][None/?x₀ … ?xₙ] e) lₖ:vₖ` with k ≠ s

I am not sure however if we want to implement this more complex scheme. My feeling is that the increase of complexity should be commensurate to the quality of life improvement.

**Comment by thedufer** [comment]
In your example it's not clear to me why we expect `h f` and `h g` to have the same behavior. Is there anything you'd recommend reading to understand these required untyped semantics?

I think I agree that this more complex scheme is probably not worth implementing. The specific problem I was hoping we could solve is one where adding optional arguments to certain polymorphic functions causes some usages of them to break. For example, `Base`'s `Option.value_exn` can't be used as the second argument to operators like `|>` or monadic `>>|` whereas stdlib's equivalent `Option.get` can. However, I think this more complex scheme would preserve that behavior, because when the expected type has a type variable as the return type, we'd have to avoid the coercion in case that type variable ends up unifying with a type containing one of the erased optional labels. Does that sound right?

**Comment by Octachron** [comment]
> In your example it's not clear to me why we expect h f and h g to have the same behavior. 

The transformation `type_argument` is transforming `h f` into `h g`, it is only valid if the two functions have the same behaviour.

Concerning your question, yes if a type variable appears in the return type of the functional argument, then one should consider that it may contain any labels and thus coercing the function would be an error.
For instance, consider:
```ocaml
let id (f: 'a -> unit ->  'a) = f

let print_and_id ?(s="default") a () =
  Format.printf "x=%s@." s;
  a

let noop ?s () = ()

let () =
  let g = id print_and_id in
  g ?s:(Some "x argument") noop () ()
```
then the optional argument `?s` of `g` must be sent as the first optional argument of `print_and_id`, and not
as the optional argument of `noop`.

Note that a possible intermediary solution would be to relax the condition from "no-labelled arguments" in the expected type to "a first non-labeled arguments and no optional arguments".
